pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,380 @@
+								LocalObjectReference: corev1.LocalObjectReference{Name: batch.Name},
+							},
+						},
+					}},
+				},
+			},
+		},
+	}
+
+	_, err = controllerutil.CreateOrUpdate(ctx, r.client, job, func() error {
+		return controllerutil.SetControllerReference(batch, job, r.scheme)
+	})
+
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("Unable to create batch job '%s': %w", batch.Name, err)
+	}
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchRunning, nil)
+}
+
+func (r *batchResource) waitToComplete() (reconcile.Result, error) {
+	batch := r.instance
+	job := &batchv1.Job{}
+	jobKey := types.NamespacedName{
+		Name:      batch.Name,
+		Namespace: batch.Namespace,
+	}
+	if err := r.client.Get(ctx, jobKey, job); err != nil {
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, fmt.Errorf("Job '%s' not found", jobKey.Name)
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, fmt.Errorf("Unable to fetch Job '%s': %w", jobKey.Name, err)
+	}
+
+	status := job.Status
+	if status.Succeeded > 0 {
+		return reconcile.Result{}, r.UpdatePhase(v2.BatchSucceeded, nil)
+	}
+
+	if status.Failed > 0 {
+		numConditions := len(status.Conditions)
+		if numConditions > 0 {
+			condition := status.Conditions[numConditions-1]
+
+			if condition.Type == batchv1.JobFailed {
+				var reason string
+				podName, err := r.getJobPodName(batch.Name, batch.Namespace)
+				if err != nil {
+					reason = err.Error()
+				} else {
+					reason, err = kubernetes.Logs(podName, batch.Namespace)
+					if err != nil {
+						reason = fmt.Errorf("Unable to retrive logs for batch %s: %w", batch.Name, err).Error()
+					}
+				}
+
+				_, err = r.update(func() {
+					r.instance.Status.Phase = v2.BatchFailed
+					r.instance.Status.Reason = reason
+				})
+				return reconcile.Result{}, err
+			}
+		}
+	}
+	// The job has not completed, wait 1 second before retrying
+	return reconcile.Result{RequeueAfter: time.Second}, nil
+}
+
+func (r *batchResource) cleanup() (reconcile.Result, error) {
+	batch := r.instance
+	meta := metav1.ObjectMeta{
+		Namespace: batch.Namespace,
+		Name:      batch.Name,
+	}
+	var allErrors error
+	// Delete the Job used to execute the batch
+	if err := r.client.Delete(ctx, &batchv1.Job{ObjectMeta: meta}); err != nil && !errors.IsNotFound(err) {
+		allErrors = fmt.Errorf("Unable to delete zero-capacity pod: %w", err)
+	}
+
+	// Delete the failed job pods used to execute the batch.
+	// Manual deletion is required as we use RestartPolicyNever on the job so that we can retrieve the cli logs
+	// in the event of a failure
+	if batch.Status.Phase == v2.BatchFailed {
+		labels := batchLabels(batch.Name)
+		r.client.DeleteAllOf(ctx, &corev1.Pod{}, []client.DeleteAllOfOption{
+			client.InNamespace(batch.Namespace),
+			client.MatchingLabels(labels),
+		}...)
+	}
+	return reconcile.Result{}, allErrors
+}
+
+func (r *batchResource) loadCluster() (*v1.Infinispan, error) {
```

Comment:
@ryanemerson do we need to validate that Infinispan CR is well formed for Batch CR execution?

Reply:
Definitely! Good catch :slightly_smiling_face:. Updated.